### ISSUE TITLE
ReQL: distributed file system

### ISSUE THREAD

**Issue by coffeemug** [rethinkdb/rethinkdb#2647]
#137 has largely been about debating different distributed file system designs, so I'd like to open a new issue to discuss a specific design:
- Users need to be able to shard individual files, which means we'll have to build this feature on top of tables.
- Users need to be able to insert, list, and delete files as first class objects.
- We should try and allow linking files from documents and to be able to specify and query user-defined metadata.

@mlucy -- could you write up a proposal?

**Comment by llambda** [comment]
Sorry if this is the wrong place, I am a rethinkdb user, first off, thanks for writing this amazing software.

I was the guy who ran into the "Error: write EPIPE when inserting 54MB file" issue above, so just wanted to chime in on this thread that as far as files or blobs are concerned, at least for me myself, my main concern would be being able to stream the binary data (from a node.js / io.js ) perspective into and out of the rethink database.

For example, consider a hypothetical 200MB video file served by a web server. I'd like to be able to stream out this video from some rethinkdb entry in a table, through the web server to the browser using streams so that it can begin playback right away. Right now I use a filesystem to stream to/from and store path strings in rethink. Thanks for listening.

**Comment by marshall007** [comment]
> I suppose this is achievable already, now by storing data in chunks in separate documents.

You could also use `.slice()` to chunk the response and then convert the cursor to a stream on the client-side. You'll still probably run into the query size limit trying to persist large files, though. It would definitely be nice to just be able to query any binary field in a document as a stream.

```js
r.http('http://httpbin.org/image/png')
  .do(function (blob) {
    var parts = blob.count().div(1024).ceil(); // ceil available in v2.1 release
    return r.range(parts).map(function (part) {
      return blob.slice(part.mul(1024), part.add(1).mul(1024))
    });
  })
```

**Comment by danielmewes** [comment]
I agree that large documents should definitely be streamable once we add proper support for them.

At the moment I think either manually splitting them up into smaller chunks/documents or using `slice` like @marshall007 suggested are the only options with RethinkDB.

**Comment by nkev** [comment]
To keep it simple, how about just storing the files where they are most at home, in the OS file system but storing meta-data for them in a RethinkDB system table so that from ReQL it feels like they are inside the database? 

For replication, a separate file watcher micro-process launched by RethinkDB on start-up could perhaps copy new/modified files across to peer servers (maybe with compression, except for image and music file types). 

It shouldn't be too hard to find an existing library to efficiently sync file changes between servers. The rsync protocol comes to mind. 

Sorry, just thinking out loud... :)

**Comment by danielmewes** [comment]
@nkev my feeling is that building a completely decoupled system for file synchronization might add more edge cases that it solves, but I haven't thought it through completely yet. We'll keep that option in mind. I guess it would allow us to take some shortcuts.

**Comment by usergit** [comment]
what's the status on this? any updates

**Comment by bchavez** [comment]
@usergit I don't know if this is what you're looking for but as far as I know there are two drivers that support a concept called **ReGrid**. **ReGrid** is analogous to Mongo DB's **GridFS**.

##### C# Driver
https://github.com/bchavez/RethinkDb.Driver/wiki/ReGrid-File-Storage

##### NodeJS:
https://github.com/internalfx/regrid

##### ReGrid Specification:
https://github.com/internalfx/regrid-spec

**Comment by nkev** [comment]
@bchavez ReGrid looks great. I like the "file system" notification events via change feeds. @danielmewes have you looked at ReGrid?

**Comment by danielmewes** [comment]
@nkev Yes. I'm actually hoping to add support for ReGrid to our official drivers eventually.
